src/cache/cache-manager.ts

```diff
@@ -69,49 +70,135 @@ export class CacheManager {
+    }
 
-      if (matchedKey) {
-        wrappedEntry = await this.cache.get<WrappedCacheEntry>(matchedKey);
-      }
+    const keys = await this.getCacheKeys();
+
+    if (!keys) return;
+
+    const siblingToken = await this.getSiblingToken(
+      cacheKey,
+      keys,
+      options.expiryAdjustmentSeconds,
+    );
+
+    if (siblingToken) {
+      return siblingToken.body;
     }
 
-    // If we still don't have an entry, exit.
-    if (!wrappedEntry) {
-      return;
+    return;
+  }
+
+  async onNoRefreshableToken(cacheKey: CacheKey): Promise<undefined> {
+    await this.cache.remove(cacheKey.toKey());
+    await this.keyManifest?.remove(cacheKey.toKey());
+    return;
+  }
+
+  async getRefreshToken(entry: WrappedCacheEntry, cacheKey: CacheKey): Promise<WrappedCacheEntry | undefined> {
+    if (!entry.body.refresh_token) {
+      return this.onNoRefreshableToken(cacheKey);
     }
 
-    const now = await this.nowProvider();
-    const nowSeconds = Math.floor(now / 1000);
+    entry.body = {
+      refresh_token: entry.body.refresh_token,
+    };
 
-    if (wrappedEntry.expiresAt - expiryAdjustmentSeconds < nowSeconds) {
-      if (wrappedEntry.body.refresh_token) {
-        wrappedEntry.body = {
-          refresh_token: wrappedEntry.body.refresh_token
-        };
+    await this.cache.set(cacheKey.toKey(), entry);
 
-        await this.cache.set(cacheKey.toKey(), wrappedEntry);
-        return wrappedEntry.body;
-      }
+    return entry;
+  };
 
-      await this.cache.remove(cacheKey.toKey());
-      await this.keyManifest?.remove(cacheKey.toKey());
+  async getActiveToken(
+    cacheKey: CacheKey,
+    expiryAdjustmentSeconds: number,
+  ): Promise<WrappedCacheEntry | undefined> {
+    const entry = await this.cache.get<WrappedCacheEntry>(
+      cacheKey.toKey()
+    );
 
-      return;
+    if (!entry) {
+      return undefined;
     }
 
-    return wrappedEntry.body;
+    const isExpired = await CacheManagerUtils.isTokenExpired(
+      entry,
+      expiryAdjustmentSeconds,
+      this.nowProvider,
+    );
+
+    return isExpired ? undefined : entry;
+  }
+
+  async getInactiveToken(
+    cacheKey: CacheKey,
+    expiryAdjustmentSeconds: number
+  ): Promise<WrappedCacheEntry | undefined> {
+    const entry = await this.cache.get<WrappedCacheEntry>(
+      cacheKey.toKey()
+    );
+
+    if (!entry) return undefined;
+
+    const isExpired = await CacheManagerUtils.isTokenExpired(
+      entry,
+      expiryAdjustmentSeconds,
+      this.nowProvider
+    );
+
+    if (isExpired) {
+      return this.getRefreshToken(entry, cacheKey);
+    }
+
+    return entry;
+  }
+
+  async getSiblingToken(
+    keyToMatch: CacheKey,
+    keys: string[],
+    expiryAdjustmentSeconds: number,
+  ): Promise<WrappedCacheEntry | undefined> {
+    const foundKey = CacheManagerUtils.findKey(keys, keyToMatch);
+
+    if (!foundKey) return undefined;
+
+    const entry = await this.cache.get<WrappedCacheEntry>(foundKey);
```

Comment:
From this line, this feels to be identical to `getInactiveToken`, which would be the 3rd time doing the exact same code.